kubernetes/cluster-dump-restore.md

```diff
@@ -0,0 +1,91 @@
+
+	If the *target cluster* has other pre-existing state, more precautions must be taken to ensure there are no undesired interactions between pre-existing resources on the *target cluster* and the dump coming from *source cluster*.
+
+	Name conflicts are easy to defect, but more subtle issues like services selecting unintended pre-existing pods is another more subtle situation that can arise. We will assume **blank target slate** for the remainder of this document.
+
+As of now, this is not entirely supported upstream. This `get --export` flag is supposed to do something like this, but leaves plenty of state that we do not consider to be portable in this case. (TODO: link to issue).
+
+We'll explain step-by-step how the CoreOS infrastructure team migrates deployments between Kubernets clusters today.
+
+### Dump from the source cluster
+
+Create a directory to hold your dump files:
+
+```shell
+mkdir ./cluster-dump
+```
+
+
+Get a list all namespaces that are not `kube-system` or `default`. This represents the list of namespaces that we want to migrate:
+
+```shell
+kubectl get --export -o=json ns | \
+jq '.items[] |
+	select(.metadata.name!="kube-system") |
+	select(.metadata.name!="default") |
+	del(.status,
+        .metadata.uid,
+        .metadata.selfLink,
+        .metadata.resourceVersion,
+        .metadata.creationTimestamp,
+        .metadata.generation
+    )' > ./cluster-dump/ns.json
+```
+
+
+For each of these namespaces, dump all services, replicationcontrollers, secrets and daemonsets. Strip any non-portable fields from the objects. If you wish to migrate additional controller resource types (replicasets, deplopyments, etc), make sure to add them to the resource type list:
+
+```shell
+for ns in $(jq -r '.metadata.name' < ./cluster-dump/ns.json);do
+    echo "Namespace: $ns"
+    kubectl --namespace="${ns}" get --export -o=json svc,rc,secrets,ds | \
+    jq '.items[] |
+        select(.type!="kubernetes.io/service-account-token") |
+        del(
+            .spec.clusterIP,
+            .metadata.uid,
+            .metadata.selfLink,
+            .metadata.resourceVersion,
+            .metadata.creationTimestamp,
+            .metadata.generation,
+            .status,
+            .spec.template.spec.securityContext,
+            .spec.template.spec.dnsPolicy,
+            .spec.template.spec.terminationGracePeriodSeconds,
+            .spec.template.spec.restartPolicy
+        )' >> "./cluster-dump/cluster-dump.json"
+done
+```
+
+Notice that pods and service tokens are explicitly omitted altogether, as they are inherently non-portable resources that are created and managed by other components. The general rule for what is portable across heterogenous clusters is is "services, controllers and secrets that aren't service tokens".
```

Comment:
For service tokens: `select(.type!="kubernetes.io/service-account-token")`

For pods, it's just that the `pod` resource type is omitted.